tower-balance/src/p2c/service.rs

```diff
@@ -230,45 +287,58 @@ where
         // previously-selected `ready_index` if appropriate.
         self.poll_discover()?;
 
-        if let Some(index) = self.ready_index {
-            debug_assert!(index < self.endpoints.len());
-            // Ensure the selected endpoint is still ready.
-            match self.poll_endpoint_index_load(index) {
-                Ok(Async::Ready(_)) => return Ok(Async::Ready(())),
-                Ok(Async::NotReady) => {}
-                Err(e) => {
-                    drop(self.endpoints.swap_remove_index(index));
-                    info!("evicting failed endpoint: {}", e.into());
-                }
-            }
-
-            self.ready_index = None;
-        }
+        self.poll_unready();
+        debug!("ready={}; unready={}", self.ready.len(), self.unready.len());
 
-        let tries = match self.endpoints.len() {
-            0 => return Ok(Async::NotReady),
-            1 => 1,
-            n => n / 2,
-        };
-        for _ in 0..tries {
-            if let Async::Ready(idx) = self.poll_ready_index().map_err(Into::into)? {
-                trace!("ready: {:?}", idx);
-                self.ready_index = Some(idx);
+        if let Some(index) = self.next_ready_index {
+            trace!("preselected ready_index={}", index);
+            debug_assert!(index < self.ready.len());
+            // Ensure the selected endpoint is still ready.
```

Comment:
Are we expecting endpoints to _stop_ being ready after they have indicated that they are ready? /cc @carllerche

Reply:
Oh, this is for cancellation, isn't it? I think it'd be useful to document somewhere exactly how this flow works. In particular, this code wouldn't be sufficient _in general_ if services could stop being ready, but it is sufficient _in this case_ because we know that only we (`&mut self`) could have cancelled the `Service`.

Reply:
@jonhoo It's possible, for instance if a failure detector or circuit breaker kicks in, for a service to go from ready to notready or ready to error. We take a belt-and-suspenders approach to avoid on the assumption that redundant polling will be cheap.

Reply:
Right, I guess my point is more that if a service can go from ready to not ready at any time, then no amount of calling `poll_ready` is sufficient. The service could always become not ready _just_ after you last call `poll_ready`. At some point, you have to leave it to `call` to actually give out the error. From this I think it follows that we are here specifically trying to deal with the case of evicted services, because for those we _know_ that this check is sufficient. This in turn makes me wonder whether we could have a fast-path here where we only do this second `poll_ready` if there has been at least one eviction?

Reply:
While it's always technically true that no amount of polling can _guarantee_ readiness, I do think it's important that the target service returns ready immediately before _call_ is invoked. so that the balancer task is not interrupted between the readiness and the call.

If we're concerned about the cost of this, let's craft some benchmarks and optimize in followups.